pandas/core/arrays/datetimes.py

```diff
@@ -197,36 +228,98 @@ class DatetimeArrayMixin(dtl.DatetimeLikeArrayMixin,
+            values = values._values
+
+        if isinstance(values, type(self)):
+            # validation
+            dtz = getattr(dtype, 'tz', None)
+            if dtz and values.tz is None:
+                dtype = DatetimeTZDtype(tz=dtype.tz)
+            elif dtz and values.tz:
+                if not timezones.tz_compare(dtz, values.tz):
+                    msg = (
+                        "Timezone of the array and 'dtype' do not match. "
+                        "'{}' != '{}'"
+                    )
+                    raise TypeError(msg.format(dtz, values.tz))
+            elif values.tz:
+                dtype = values.dtype
+            # freq = validate_values_freq(values, freq)
+            if freq is None:
+                freq = values.freq
+            values = values._data
+
+        if not isinstance(values, np.ndarray):
+            msg = (
+                "Unexpected type '{}'. 'values' must be a DatetimeArray "
+                "ndarray, or Series or Index containing one of those."
+            )
+            raise ValueError(msg.format(type(values).__name__))
+
         if values.dtype == 'i8':
             # for compat with datetime/timedelta/period shared methods,
             #  we can sometimes get here with int64 values.  These represent
             #  nanosecond UTC (or tz-naive) unix timestamps
-            values = values.view('M8[ns]')
+            values = values.view(_NS_DTYPE)
 
-        assert values.dtype == 'M8[ns]', values.dtype
+        if values.dtype != _NS_DTYPE:
+            msg = (
+                "The dtype of 'values' is incorrect. Must be 'datetime64[ns]'."
+                " Got {} instead."
+            )
+            raise ValueError(msg.format(values.dtype))
+
+        # Performance note:
+        # for a length 10,000 ndarray[datetime64[ns]], pandas_dtype() takes
+        # ~15% of the runtime of __init__. It's only useful for converting
+        # string aliases like 'M8[ns]' or 'datetime64[ns, tz]'.
+        # We should consider requiring an actual dtype.
+
+        dtype = pandas_dtype(dtype)
+        if (isinstance(dtype, np.dtype) and dtype != _NS_DTYPE
+                or not isinstance(dtype, (np.dtype, DatetimeTZDtype))):
+            msg = (
+                "Unexpected value for 'dtype': '{}'. "
+                "Must be 'datetime64[ns]' or DatetimeTZDtype'."
+            )
+            raise ValueError(msg.format(dtype))
 
-        result = object.__new__(cls)
-        result._data = values
-        result._freq = freq
-        tz = timezones.maybe_get_tz(tz)
-        result._tz = timezones.tz_standardize(tz)
-        return result
+        if freq == "infer":
+            msg = (
+                "Frequency inference not allowed in DatetimeArray.__init__. "
+                "Use 'pd.array()' instead."
+            )
+            raise ValueError(msg)
+
+        if copy:
+            values = values.copy()
+        if freq:
+            freq = to_offset(freq)
```

Comment:
ditto re not doing freq validation

Reply:
https://github.com/pandas-dev/pandas/pull/24024/files#r243105181

There is a request-for-fleshing-out there that I'll get to now.